src/ralph/backends/data/async_es.py

```diff
@@ -0,0 +1,407 @@
+                If chunk_size is `None` it defaults to `DEFAULT_CHUNK_SIZE`.
+            ignore_errors (bool): If `True`, errors during the write operation
+                will be ignored and logged. If `False` (default), a `BackendException`
+                will be raised if an error occurs.
+            operation_type (BaseOperationType or None): The mode of the write operation.
+                If `operation_type` is `None`, the `default_operation_type` is used
+                instead. See `BaseOperationType`.
+
+        Return:
+            int: The number of written documents.
+
+        Raise:
+            BackendException: If a failure occurs while writing to Elasticsearch or
+                during document decoding and `ignore_errors` is set to `False`.
+            BackendParameterException: If the `operation_type` is `APPEND` as it is not
+                supported.
+        """
+        count = 0
+        data = iter(data)
+        try:
+            first_record = next(data)
+        except StopIteration:
+            logger.info("Data Iterator is empty; skipping write to target.")
+            return count
+        if not operation_type:
+            operation_type = self.default_operation_type
+        target = target if target else self.settings.DEFAULT_INDEX
+        chunk_size = chunk_size if chunk_size else self.settings.DEFAULT_CHUNK_SIZE
+        if operation_type == BaseOperationType.APPEND:
+            msg = "Append operation_type is not supported."
+            logger.error(msg)
+            raise BackendParameterException(msg)
+
+        data = chain((first_record,), data)
+        if isinstance(first_record, bytes):
+            data = self._parse_bytes_to_dict(data, ignore_errors)
+
+        logger.debug(
+            "Start writing to the %s index (chunk size: %d)", target, chunk_size
+        )
+        try:
+            async for success, action in async_streaming_bulk(
+                client=self.client,
+                actions=self._to_documents(data, target, operation_type),
+                chunk_size=chunk_size,
+                raise_on_error=(not ignore_errors),
+                refresh=self.settings.REFRESH_AFTER_WRITE,
+            ):
+                count += success
+                logger.debug("Wrote %d document [action: %s]", success, action)
+
+            logger.info("Finished writing %d documents with success", count)
+        except (BulkIndexError, ApiError, TransportError) as error:
+            msg = "%s %s Total succeeded writes: %s"
+            details = getattr(error, "errors", "")
+            logger.error(msg, error, details, count)
+            raise BackendException(msg % (error, details, count)) from error
+        return count
+
+    async def close(self) -> None:
+        """Wrap the AsyncElasticsearch.close method.
+
+        Raises:
+            BackendException: raised for any failure.
+        """
+        await self.client.close()
```

Comment:
Missing exception handling? Also, if `self._client` is `None` we could skip calling `self.client.close()`

Reply:
Thanks, changed!